lib/accesscontrol/general.rb

```diff
@@ -0,0 +1,30 @@
+module AccessControl
+
+  # Permission checks not directly related to records happen here.
+  module General
+    module_function
+
+    # Ask whether the actor has permission to perform action_id
+    # in the given namespace. Multiple actions can have the same id
+    # as long as their namespace is different. The namespace can be
+    # any String. We recommend using namespace to group a class of
+    # permissions, such as to group parts of a particular feature
+    # in your application.
+    #
+    # @param actor [ActiveRecord::Base] The actor we're checking for permission on.
+    # @param action_id [Integer, Array<Integer>] The action or actions we're checking whether the actor has. If this is an array, then the check is ORed.
+    # @param namespace [String] The namespace of the given action_id.
+    # @return [Boolean] Returns true if actor has been granted the permission, false otherwise.
+    #
+    # @example
+    #   # Can the user perform the action with id 3 for posts?
+    #   AccessControl::General.can?(user, 3, "posts")
+    def can?(actor, action_id, object_type)
```

Comment:
> It seems like on the scale of things, I'm least comfortable with making it a module method, slightly more comfortable if we made it a class method, but ideally it would just be an instantiated object.

🤔 - should Accessly(tm) be a singleton?

Reply:
lol on the 'tm'! 

Hmm, I'd be open to thoughts on where that might be an advantage to us... 

I think my initial thought is that I'd be more of a fan of simple short lived objects mainly because it helps eliminate any baggage that might arise from having longer persistent state that might get mutated in surprising ways? I can kind of see when you have a connection you need to configure or something how that might be worth the trade off at some point... but even in those cases, I would almost favor having an object that represents the connection configuration and pass that in when needed? Not sure... but I'd love to hear more if it seems like it would be a help here!

Reply:
> `access_check.reason_for_declined_permission`

For that specific example, I feel we should keep the gem simple. We know the reason they were declined - they were never granted that permission!

There is one state that I think makes sense to keep around, which a singleton would be ideal for: cacheing permissions. Say we ask for whether a user has permission to create widgets. Then a few lines later, we ask whether that user has permission to create foobars. We can avoid having to call out to the database twice if (in the gem) we just find out all the things the user can do and cacheing that so subsequent requests would be findable in memory.

If we went the all-module-methods route, this would represent the functional programming model. We ask `AccessControl` to do something or answer a question, the thing happens, and no state is left behind. I'm definitely a fan of this method as I believe it helps make the code path clear and doesn't leave much room for stateful baggage, memory leaks, etc.

As for my opinion, I like the idea of the external API being module methods (like `AccessControl.can?` like we have implemented now), and having those module methods call out to a singleton which does the work, including cacheing.

I like the API module methods because it keeps the API simple and easy to understand. From our app, we don't need to know instantiate or store anything - we just ask "can this user do this thing?" and the gem does its magic to tell me.

Reply:
Great points! 

That is a good point with the singleton... might be a good way to add a caching layer. I wonder though if that could become a potential point for memory bloat if over time we are holding all those things... I guess you could do a timed cache. But then you would also have to think about busting the cache for when permissions are updated 🤔 seems like it could add complexity, but could be a tool that we could use if we find that is a bottleneck? I think that we get a degree of query caching through active record as well right?

Reply:
When I talk about caching, I mainly just mean something that lives for a single request. If I look up permissions once during a request, then I shouldn't have to hit the database to get those same permissions during the same request.

However, when I hit the server with a new request, I should be going to the database again.

That way, we don't have to worry about cache invalidation!